File: src/cli/connection/update.ts
import { Command, ExactArgs, ref } from "vipvot";
import { updateConnection, getCredential, getCredentials } from "../../lib/config.ts";
import { printError, printJsonRaw } from "../../lib/output.ts";

export function buildUpdateCommand(): Command {
  const credential = ref<string>("");
  const clearCredential = ref<boolean>(false);
  const database = ref<string>("");

  const cmd = Command({
    use: "update <alias>",
    short: "Update a saved connection",
    args: ExactArgs(1),
    run: (_cmd, args) => {
      const [alias] = args as [string];
      try {
        if (credential.value && clearCredential.value) {
          throw new Error("Cannot use --credential and --clear-credential together.");
        }

        if (credential.value) {
          const cred = getCredential(credential.value);
          if (!cred) {
            const available = Object.keys(getCredentials());
            throw new Error(
              `Credential "${credential.value}" not found. Available: ${available.join(", ") || "(none)"}. Run: agent-mongo credential add <alias> --username <user> --password <pass>`,
            );
          }
        }

        const updates: Record<string, string | undefined> = {};
        if (database.value) {
          updates.database = database.value;
        }
        if (clearCredential.value) {
          updates.credential = undefined;
        } else if (credential.value) {
          updates.credential = credential.value;
        }

        updateConnection(alias, updates);
        printJsonRaw({ ok: true, alias, updated: Object.keys(updates) });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Failed to update connection");
      }
    },
  });

  cmd.flags().stringVarP(credential, "credential", "", "", "Credential alias for authentication");
  cmd
    .flags()
    .boolVarP(clearCredential, "clear-credential", "", false, "Remove credential from connection");
  cmd.flags().stringVarP(database, "database", "", "", "Override database name");
  cmd.markFlagsMutuallyExclusive("credential", "clear-credential");

  return cmd;
}

File: internal/cli/connection/update.go
package connection

import (
	"github.com/spf13/cobra"

	"github.com/shhac/agent-mongo/internal/config"
	"github.com/shhac/agent-mongo/internal/credential"
	"github.com/shhac/agent-mongo/internal/output"
)

func registerUpdate(parent *cobra.Command) {
	var credentialAlias, database string
	var clearCredential bool

	cmd := &cobra.Command{
		Use:   "update <alias>",
		Short: "Update a saved connection",
		Args:  cobra.ExactArgs(1),
		RunE: func(_ *cobra.Command, args []string) error {
			alias := args[0]

			if credentialAlias != "" {
				if err := credential.RequireExists(credentialAlias); err != nil {
					return err
				}
			}

			var updates config.ConnectionUpdates
			var updated []string
			if database != "" {
				updates.Database = &database
				updated = append(updated, "database")
			}
			if clearCredential {
				empty := ""
				updates.Credential = &empty
				updated = append(updated, "credential")
			} else if credentialAlias != "" {
				updates.Credential = &credentialAlias
				updated = append(updated, "credential")
			}

			if err := config.UpdateConnection(alias, updates); err != nil {
				return err
			}
			return output.PrintRaw(map[string]any{"ok": true, "alias": alias, "updated": updated})
		},
	}

	cmd.Flags().StringVar(&credentialAlias, "credential", "", "Credential alias for authentication")
	cmd.Flags().BoolVar(&clearCredential, "clear-credential", false, "Remove credential from connection")
	cmd.Flags().StringVar(&database, "database", "", "Override database name")
	cmd.MarkFlagsMutuallyExclusive("credential", "clear-credential")
	parent.AddCommand(cmd)
}
